File: assign1/ProdCons1.java
import java.util.LinkedList;
import java.util.Queue;

class ProdCons1 extends Thread {
  static final int BUFSIZE = 5; 
  static final int NUMITEMS = 50; 
  static Object lck = new Object();
  static Queue<Integer> q = new LinkedList<>();

  ProdCons1() {}


  public static void main(String[] args) throws InterruptedException {
      final ProdCon pc = new ProdCon();

      Thread prod = new Thread() {
        public void run() 
        {
          try {
            pc.producer();
          } 
          catch (InterruptedException e) {
            e.printStackTrace();
          }
        }
      };
  
      Thread con = new Thread() {
        public void run() 
        {
          try {
            pc.consumer();
          } 
          catch (InterruptedException e) {
            e.printStackTrace();
          }
        }
      };

      prod.start();
      con.start();
      prod.join();
      con.join();
        
      System.out.println("Main: all done! So leave me alone!");
  }
  public static class ProdCon {
    private void producer() throws InterruptedException{ 
        int i = 0;
        System.out.println("Producer is starting...");
        while(i < NUMITEMS){
          ++i;
          synchronized(lck) {
            while(q.size() == BUFSIZE) {
                lck.wait();
            }
            System.out.println("Producer added " + i + " (qsz: " + q.size() + ")");
            q.add(i);
            lck.notify();
          }
        }
        System.out.println("Producer is ending...");
    }

    private void consumer()  throws InterruptedException{ 
        System.out.println("Consumer is starting...");
        int i = 0;
        int num;
        while(i < NUMITEMS){
          synchronized(lck) {
            while(q.size() == 0) {
                lck.wait();
            }
            num = q.remove();
            System.out.println("Consumer consumed " + num + " (qsz: " + q.size() + ")");
            lck.notify();
          }
          ++i;
        }
        System.out.println("Consumer is ending...");
    }
  }
}

File: assign1/prodcons1.cpp
/* One producer, One consumer, 3 threads 
 * Using circular Queue, a conditional variable
 * and a mutex lock */
#include "queue.h"
#include <iostream> 
#include <thread> 
#include <mutex> 
#include <condition_variable> 
using namespace std; 

mutex mtx;   
condition_variable cv;
int BUFFSIZE = 20;
int NUMITEMS = 100; 
Queue pcq(BUFFSIZE);

void producer(void) {
  int core = sched_getcpu();
  cout << "Producer starting on core " + to_string(core) + "\n";

  for(int i = 1; i <= NUMITEMS; ++i){
    unique_lock<mutex> lck(mtx);
    while(pcq.isFull()) 
    {
      cv.notify_one();
      cv.wait(lck);
    } 
    pcq.add(i);
    cv.notify_one();
    lck.unlock();
    cout << "Producer added " + to_string(i) + " (qsz: " + to_string(pcq.size()) + ")\n"; 
  }
  cv.notify_one();
  cout << "Producer ending\n";
}
  
void consumer(void) {
  int core = sched_getcpu();
  cout << "Consumer starting on core " + to_string(core) + "\n";
  for(int i = 1; i <= NUMITEMS; ++i){
    unique_lock<mutex> lck(mtx);
    while(pcq.isEmpty()) 
    {
      cv.notify_one();
      cv.wait(lck);
    } 
    pcq.remove();
    cv.notify_one();
    lck.unlock();
    cout << "Consumer removed " + to_string(i) + " (qsz: " + to_string(pcq.size()) + ")\n"; 
  }
  cv.notify_one();
  cout << "Consumer ending\n";
}

int main(int argc, char **argv) {
  thread t1, t2;
  t1 = thread(producer);
  t2 = thread(consumer);
  cv.notify_all();
  t1.join();
  t2.join();
  cout << "Main: all done! Now leave me alone!\n";
}  
